src/e2e-test/java/io/cdap/plugin/gcs/GCSValidationHelper.java

```diff
@@ -0,0 +1,224 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.gcs;
+
+import au.com.bytecode.opencsv.CSVReader;
+import com.google.api.gax.paging.Page;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageOptions;
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import io.cdap.e2e.utils.PluginPropertyUtils;
+import org.apache.avro.file.DataFileReader;
+import org.apache.avro.file.FileReader;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DatumReader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * GCS Validation Helper.
+ */
+public class GCSValidationHelper {
+  private static final String avroFilePath = PluginPropertyUtils.pluginProp("gcsAvroExpectedFilePath");
+  private static final String csvFilePath = PluginPropertyUtils.pluginProp("gcsCsvExpectedFilePath");
+  private static final String projectId = PluginPropertyUtils.pluginProp("projectId");
+  private static final Gson gson = new Gson();
+  private static final Logger LOG = LoggerFactory.getLogger(GCSValidationHelper.class);
+
+  /**
+   * Validates if the data in a (GCS) bucket matches the expected CSV data in JSON format.
+   *
+   * @param bucketName The name of the GCS bucket to validate.
+   * @return True if the GCS CSV data matches the expected data, false otherwise.
+   * @throws IOException If an IO error occurs during data retrieval.
+   */
+  public static boolean validateGCSSourceToGCSSinkWithCSVFormat(String bucketName) {
+    Map<String, JsonObject> expectedCSVData = readCsvAndConvertToJson();
+    Map<String, JsonObject> actualGcsCsvData = listBucketObjects(bucketName, DataFormat.CSV);
+
+    boolean isMatched = actualGcsCsvData.equals(expectedCSVData);
+
+    return isMatched;
+  }
+
+  /**
+   * Validates if the data in a (GCS) bucket matches the data
+   * obtained from Avro files converted to JSON.
+   *
+   * @param bucketName The name of the GCS bucket to validate.
+   * @return True if the GCS data matches the Avro data, false otherwise.
+   * @throws IOException If an IO error occurs during data retrieval.
+   */
+  public static boolean validateGCSSourceToGCSSinkWithAVROFormat(String bucketName) throws IOException {
```

Comment:
This is unused rn are we planning to use it later ?

Reply:
Yes, this function is being used in all data types scenarios where the file format is in Avro. Restructured and make a generic validation helper class for the parsing of csv, avro, json file formats.